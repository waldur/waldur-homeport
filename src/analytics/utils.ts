import * as moment from 'moment-timezone';

import { dictToList, formatFilesize } from '@waldur/core/utils';
import { translate } from '@waldur/i18n';

import * as constants from './constants';
import { getRegisteredQuota } from './registry';

export const isQuotaExceeds = quota =>
  quota.limit >= 0 && quota.limit <= quota.usage;

export const sortObjecstByProp = (propName: string) => (
  objects: Array<{ [key: string]: string }>,
  locale?: string | string[],
  options?: { [key: string]: string | boolean },
) =>
  [...objects].sort((a, b) =>
    a[propName].localeCompare(b[propName], locale, options),
  );

export const sortQuotasByLabel = sortObjecstByProp('label');

export const sortProjectsByName = sortObjecstByProp('name');

export const quotasRegitryFilter = quotas =>
  quotas.filter(quota => !!getRegisteredQuota(quota.name));

export const getMaxFileSizeNames = quotas => {
  const fileSizeNames = {};
  for (const quota of quotas) {
    const registeredQuotaConfig = getRegisteredQuota(quota.name);
    if (
      (registeredQuotaConfig &&
        registeredQuotaConfig.valueType !== constants.valueType.byte) ||
      !registeredQuotaConfig
    ) {
      continue;
    }
    const size = quota.limit === -1 ? quota.usage : quota.limit;
    fileSizeNames[quota.name] =
      size === 0
        ? 'GB'
        : formatFilesize(size)
            .slice(-2)
            .trim();
  }
  return fileSizeNames;
};

export const addRegistryConfig = quotas =>
  quotas.map(quota => {
    const {
      formatterTemplateBar = null,
      formatterTemplatePie = null,
      label = '',
    } = getRegisteredQuota(quota.name) ? getRegisteredQuota(quota.name) : {};
    return {
      ...quota,
      formatterTemplateBar,
      formatterTemplatePie,
      label,
    };
  });

export const formatQuotaUsage = quota => {
  const formatter = getRegisteredQuota(quota.name).formatter;
  return formatter
    ? {
        ...quota,
        usage:
          quota.usage > 0 ? Math.round(formatter(quota.usage)) : quota.usage,
        limit:
          quota.limit !== -1 ? Math.round(formatter(quota.limit)) : quota.limit,
      }
    : quota;
};

export const setHistoryQuotasName = quotas =>
  quotas.map(quota => {
    const { data = [] } = quota;
    for (const item of data) {
      if (item.object && item.object.name) {
        return {
          ...quota,
          name: item.object.name,
        };
      }
    }
    return quota;
  });

export const getPieChartsData = quotas =>
  quotas.map(quota => ({
    id: quota.uuid[0],
    limit: quota.limit,
    label: quota.label,
    maxFileSizeName: quota.maxFileSizeName,
    exceeds: isQuotaExceeds(quota),
    options: {
      color: [constants.chartColors.orange, constants.chartColors.blue],
      title: {
        text: quota.label,
      },
      tooltip: {
        formatter:
          typeof quota.formatterTemplatePie === 'function'
            ? quota.formatterTemplatePie(quota.maxFileSizeName)
            : null,
      },
      series: [
        {
          name: quota.label,
          type: 'pie',
          legendHoverLink: false,
          hoverAnimation: false,
          hoverOffset: 0,
          clockwise: false,
          labelLine: {
            show: false,
          },
          data: [
            {
              name: translate('Usage'),
              value: quota.usage,
            },
            {
              name: translate('Remaining limit'),
              value: quota.limit > 0 ? quota.limit - quota.usage : 0,
            },
          ],
        },
      ],
    },
  }));

export const getBarChartsData = quotas =>
  quotas.map(quota => {
    if (!quota.data) {
      return quota;
    }
    const dateData = [];
    const limitData = [];
    const usageData = [];
    for (const item of quota.data) {
      const { point, object: { limit, usage } = { limit: 0, usage: 0 } } = item;
      let limitDataItem = limit === -1 ? 0 : limit;
      let usageDataItem = usage;
      if (typeof quota.formatterTemplateBar === 'function') {
        limitDataItem = parseFloat(
          formatFilesize(
            limitDataItem,
            quota.minFileSizeName,
            quota.maxFileSizeName,
          ),
        );
        usageDataItem = parseFloat(
          formatFilesize(usage, quota.minFileSizeName, quota.maxFileSizeName),
        );
      }
      dateData.push(moment(point * 1000).format('Do MMMM'));
      limitData.push(limitDataItem);
      usageData.push(usageDataItem);
    }
    return {
      id: quota.uuid[0],
      label: quota.label,
      loading: quota.loading,
      erred: quota.erred,
      exceeds: isQuotaExceeds(quota.data.slice(-1)[0].object),
      maxFileSizeName: quota.maxFileSizeName,
      options: {
        color: [constants.chartColors.orange, constants.chartColors.blue],
        title: {
          text: quota.label,
        },
        tooltip: {
          trigger: 'axis',
          axisPointer: {
            type: 'shadow',
          },
          formatter:
            typeof quota.formatterTemplateBar === 'function'
              ? quota.formatterTemplateBar(quota.maxFileSizeName)
              : null,
        },
        toolbox: {
          show: true,
          orient: 'vertical',
          left: 'right',
          top: 'center',
          feature: {
            mark: { show: true },
            dataView: { show: true, readOnly: false },
            magicType: { show: true, type: ['line', 'bar', 'stack', 'tiled'] },
            restore: { show: true },
            saveAsImage: { show: true },
          },
        },
        calculable: true,
        xAxis: [
          {
            type: 'category',
            axisTick: { show: false },
            data: dateData,
          },
        ],
        yAxis: [
          {
            type: 'value',
          },
        ],
        series: [
          {
            name: translate('Usage'),
            type: 'bar',
            barGap: 0,
            data: usageData,
          },
          {
            name: translate('Limit'),
            type: 'bar',
            barGap: 0,
            data: limitData,
          },
        ],
      },
    };
  });

export const getExceededQuotas = quotas => {
  const exceededQuotas = [];
  for (const quota of quotas) {
    if (!isQuotaExceeds(quota)) {
      continue;
    }
    exceededQuotas.push(quota);
  }
  return exceededQuotas;
};

export const formatFilter = vmFilter => {
  const vmFilterCopy = { ...vmFilter };
  if (vmFilter && vmFilter.service_provider) {
    vmFilterCopy.service_provider = vmFilter.service_provider.map(
      provider => provider.value,
    );
    return vmFilterCopy;
  }
  return vmFilter;
};

export const formatServiceProviders = serviceProviders => {
  return serviceProviders.map(serviceProvider => {
    return {
      value: serviceProvider.uuid,
      name: serviceProvider.name,
    };
  });
};

export const mergeQuotas = (quotaA, quotaB) => {
  if (!quotaA && !quotaB) {
    return null;
  }
  if (!quotaA) {
    return quotaB;
  }
  if (!quotaB) {
    return quotaA;
  }
  return {
    ...quotaA,
    uuid: Array.isArray(quotaA.uuid)
      ? [...quotaA.uuid, quotaB.uuid]
      : [quotaA.uuid, quotaB.uuid],
    limit:
      quotaB.limit > -1
        ? quotaA.limit > -1
          ? quotaA.limit + quotaB.limit
          : quotaB.limit
        : quotaA.limit,
    usage: quotaA.usage + quotaB.usage,
  };
};

export const combineQuotas = quotas => {
  const combinedQuotas = {};
  for (const currQuotas of quotas) {
    for (const quota of currQuotas) {
      const combinedQuota = combinedQuotas[quota.name];
      if (combinedQuota) {
        combinedQuotas[quota.name] = mergeQuotas(combinedQuota, quota);
      } else {
        combinedQuotas[quota.name] = {
          ...quota,
          uuid: [quota.uuid],
        };
      }
    }
  }
  return dictToList(combinedQuotas);
};

export const combineHistoryQuotas = quotas => {
  const combinedQuotas = {};
  for (const currQuotas of dictToList(quotas)) {
    for (const quota of dictToList(currQuotas)) {
      if (!quota || !quota.name || quota.loading) {
        return null;
      }
      const combinedQuota = combinedQuotas[quota.name];
      if (combinedQuota) {
        const data = combinedQuota.data.map((timePointQuota, index) => ({
          ...timePointQuota,
          object: mergeQuotas(timePointQuota.object, quota.data[index].object),
        }));
        combinedQuotas[quota.name] = {
          ...combinedQuota,
          uuid: [...combinedQuota.uuid, quota.uuid],
          data,
        };
      } else {
        combinedQuotas[quota.name] = {
          ...quota,
          uuid: [quota.uuid],
        };
      }
    }
  }
  return dictToList(combinedQuotas);
};
