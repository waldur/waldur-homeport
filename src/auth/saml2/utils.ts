import { getSaml2IdentityProviders } from '@waldur/auth/saml2/api';

export const fetchIdentityProviderOptions = (query: string) =>
  getSaml2IdentityProviders(query).then((options) => ({ options }));

export const redirectPost = (url: string, data: object) => {
  const form = document.createElement('form');
  document.body.appendChild(form);
  form.method = 'POST';
  form.action = url;
  for (const name in data) {
    if (data.hasOwnProperty(name)) {
      const input = document.createElement('input');
      input.type = 'hidden';
      input.name = name;
      input.value = data[name];
      form.appendChild(input);
    }
  }
  form.submit();
};
