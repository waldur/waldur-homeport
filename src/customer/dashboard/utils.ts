import { CustomerActionsProps } from './types';

export const checkPermissions = (props: CustomerActionsProps) => {
  const isStaff = props.user.is_staff;
  const isOwner = props.customer.owners.find(
    owner => owner.uuid === props.user.uuid,
  );
  return isStaff || isOwner;
};
