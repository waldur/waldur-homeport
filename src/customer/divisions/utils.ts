import { interpolateInferno } from 'd3-scale-chromatic';
import { DateTime } from 'luxon';

import { translate } from '@waldur/i18n';

const calculatePoint = (i, intervalSize, colorRangeInfo) => {
  const { colorStart, colorEnd, useEndAsStart } = colorRangeInfo;
  return useEndAsStart
    ? colorEnd - i * intervalSize
    : colorStart + i * intervalSize;
};

const interpolateColors = (dataLength, colorScale, colorRangeInfo) => {
  const { colorStart, colorEnd } = colorRangeInfo;
  const colorRange = colorEnd - colorStart;
  const intervalSize = colorRange / dataLength;
  let i, colorPoint;
  const colorArray = [];

  for (i = 0; i < dataLength; i++) {
    colorPoint = calculatePoint(i, intervalSize, colorRangeInfo);
    colorArray.push(colorScale(colorPoint));
  }

  return colorArray;
};

const getEChartOption = (chartData) => ({
  toolbox: {
    feature: {
      saveAsImage: {
        title: translate('Save'),
        name: `Organizations-by-divisions-chart-${DateTime.now().toISODate()}`,
        show: true,
      },
    },
  },
  series: {
    type: 'sunburst',
    highlightPolicy: 'ancestor',
    data: chartData,
    radius: [0, '90%'],
    label: {
      rotate: 'radial',
    },
  },
});

export const generateColors = (amount: number, colorRangeInfo): string[] =>
  interpolateColors(amount, interpolateInferno, colorRangeInfo);

const buildDivisionsHierarchy = (array) => {
  const COLORS = generateColors(array.length, {
    colorStart: 0.2,
    colorEnd: 0.8,
    useEndAsStart: false,
  });

  const roots = [],
    children = {};

  // find the top level nodes and hash the children based on parent
  for (let i = 0, len = array.length; i < len; ++i) {
    const item = array[i],
      p = item.parent_uuid,
      target = !p ? roots : children[p] || (children[p] = []);
    target.push({
      ...item,
      itemStyle: {
        color: COLORS[i],
      },
    });
  }

  // function to recursively build the tree
  const findChildren = (parent) => {
    if (children[parent.uuid]) {
      parent.children = children[parent.uuid];
      for (let i = 0, len = parent.children.length; i < len; ++i) {
        findChildren(parent.children[i]);
      }
    }
  };

  // enumerate through to handle the case where there are multiple roots
  for (let i = 0, len = roots.length; i < len; ++i) {
    findChildren(roots[i]);
  }

  return roots;
};

const addCustomersWithoutDivision = (data, customers) => {
  const totalNumberOfCustomersWithoutDivision = customers.reduce(
    (acc, customer) => (!customer.division_uuid ? acc + 1 : acc),
    0,
  );
  data.push({
    name: 'Uncategorized',
    value: totalNumberOfCustomersWithoutDivision,
    itemStyle: {
      color: '#C0C0C0',
    },
  });
};

const addValueToDivisions = (divisions, customers) =>
  divisions.map((division) => {
    const countOccurrences = customers.reduce(
      (acc, customer) =>
        customer.division_uuid === division.uuid ? acc + 1 : acc,
      0,
    );
    return { ...division, value: countOccurrences };
  });

export const getEChartOptions = ({ divisions, customers }) => {
  divisions = addValueToDivisions(divisions, customers);
  const data = buildDivisionsHierarchy(divisions);
  addCustomersWithoutDivision(data, customers);
  return getEChartOption(data);
};
