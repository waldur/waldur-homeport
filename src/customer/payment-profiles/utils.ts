import { translate } from '@waldur/i18n';

export const getInitialValues = (props) => ({
  name: props.resolve.name,
  payment_type: props.resolve.payment_type,
  end_date: props.resolve.attributes.end_date,
  agreement_number: props.resolve.attributes.agreement_number,
  contract_sum: props.resolve.attributes.contract_sum,
});

export const getPaymentProfileTypeOptions = () => [
  { label: translate('Fixed-price contract'), value: 'fixed_price' },
  { label: translate('Monthly invoices'), value: 'invoices' },
  {
    label: translate('Payment gateways (monthly)'),
    value: 'payment_gw_monthly',
  },
];
