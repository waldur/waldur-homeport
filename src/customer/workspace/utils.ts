import { get } from '@waldur/core/api';
import { ngInjector } from '@waldur/core/services';
import { translate } from '@waldur/i18n';
import { MenuItemType } from '@waldur/navigation/sidebar/types';
import { Customer } from '@waldur/workspace/types';

export const getSidebarItems = (customer: Customer): MenuItemType[] => [
  {
    key: 'dashboard',
    label: translate('Dashboard'),
    icon: 'fa-th-large',
    state: 'organization.dashboard',
    params: {
      uuid: customer.uuid,
    },
    index: 100,
  },
  {
    key: 'projects',
    label: translate('Projects'),
    icon: 'fa-bookmark',
    state: 'organization.projects',
    params: {
      uuid: customer.uuid,
    },
    feature: 'projects',
    countFieldKey: 'projects',
    index: 300,
  },
  {
    key: 'events',
    label: translate('Audit logs'),
    icon: 'fa-bell-o',
    state: 'organization.details',
    params: {
      uuid: customer.uuid,
    },
    feature: 'eventlog',
    index: 600,
  },
  {
    key: 'issues',
    label: translate('Issues'),
    icon: 'fa-question-circle',
    state: 'organization.issues',
    params: {
      uuid: customer.uuid,
    },
    feature: 'support',
    index: 700,
  },
  {
    key: 'team',
    label: translate('Team'),
    icon: 'fa-group',
    state: 'organization.team',
    params: {
      uuid: customer.uuid,
    },
    feature: 'team',
    countFieldKey: 'users',
    index: 900,
  },
  {
    key: 'billing',
    label: ngInjector.get('BillingUtils').getTabTitle(),
    icon: 'fa-file-text-o',
    state: 'organization.billing.tabs',
    params: {
      uuid: customer.uuid,
    },
    feature: 'billing',
    index: 1000,
  },
  {
    key: 'manage',
    label: translate('Manage'),
    icon: 'fa-wrench',
    state: 'organization.manage',
    params: {
      uuid: customer.uuid,
    },
    index: 9999,
  },
];

export const getCustomerCounters = (customer: Customer, fields: string[]) =>
  get(`/customers/${customer.uuid}/counters/`, { params: { fields } }).then(
    response => response.data,
  );

export const getExtraSidebarItems = (): Promise<MenuItemType[]> => {
  const SidebarExtensionService = ngInjector.get('SidebarExtensionService');
  return SidebarExtensionService.getItems('customer');
};
