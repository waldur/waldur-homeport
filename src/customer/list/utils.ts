import { DateTime } from 'luxon';

import { Customer } from '@waldur/workspace/types';

const RIGHT_ARROW_STRING = '\u2192';

export const getInitialValuesOfOrganizationUpdateForm = (
  customer: Customer,
) => ({
  name: customer.name,
  uuid: customer.uuid,
  abbreviation: customer.abbreviation,
  division: {
    url: customer.division,
    name: `${customer.division_parent_name} ${RIGHT_ARROW_STRING} ${customer.division_name}`,
  },
  contact_details: customer.contact_details,
  registration_code: customer.registration_code,
  agreement_number: customer.agreement_number,
  sponsor_number: customer.sponsor_number,
  email: customer.email,
  phone_number: customer.phone_number,
  access_subnets: customer.access_subnets,
  homepage: customer.homepage,
  country: {
    value: customer.country,
    display_name: customer.country_name,
  },
  vat_code: customer.vat_code,
  default_tax_percent: customer.default_tax_percent,
});

export const makeAccountingPeriods = (start: DateTime) => {
  let date = DateTime.now().startOf('month');
  const diff = date.diff(start, 'months');
  const choices = [];
  for (let i = 0; i <= diff.months; i++) {
    const month = date.month;
    const year = date.year;
    const label = date.toFormat('MMMM, yyyy');
    choices.push({
      label,
      value: { year, month, current: i === 0 },
    });
    date = date.minus({ months: 1 });
  }
  return choices;
};
