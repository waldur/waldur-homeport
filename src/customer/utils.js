import { WOKSPACE_NAMES } from '../navigation/workspace/constants';

// @ngInject
export function loadCustomer(
  $q,
  $stateParams,
  $state,
  customersService,
  currentStateService,
  WorkspaceService,
) {
  if (!$stateParams.uuid) {
    return $q.reject();
  }
  return customersService
    .$get($stateParams.uuid)
    .then(customer => {
      currentStateService.setCustomer(customer);
      return customer;
    })
    .then(customer => {
      WorkspaceService.setWorkspace({
        customer: customer,
        project: null,
        hasCustomer: true,
        workspace: WOKSPACE_NAMES.organization,
      });
      return customer;
    })
    .catch(error => {
      if (error.status === 404) {
        $state.go('errorPage.notFound');
      }
    });
}

// @ngInject
export function CustomerController(
  $scope,
  $state,
  usersService,
  currentStateService,
  customersService,
) {
  usersService.getCurrentUser().then(currentUser => {
    currentStateService.getCustomer().then(currentCustomer => {
      $scope.currentCustomer = currentCustomer;
      $scope.currentUser = currentUser;

      if (
        customersService.checkCustomerUser(currentCustomer, currentUser) ||
        currentUser.is_support
      ) {
        currentStateService.setOwnerOrStaff(true);
      } else {
        currentStateService.setOwnerOrStaff(false);
        $state.go('profile.details');
      }
    });
  });
}
