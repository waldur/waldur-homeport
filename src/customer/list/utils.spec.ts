import { advanceTo, clear } from 'jest-date-mock';
import { DateTime } from 'luxon';

import { makeAccountingPeriods } from './utils';

afterAll(() => {
  clear();
});

describe('makeAccountingPeriods', () => {
  it('iterates through months', () => {
    advanceTo(
      DateTime.fromObject({ year: 2021, month: 10, day: 1 }).toJSDate(),
    );
    const options = makeAccountingPeriods(
      DateTime.fromObject({ year: 2021, month: 1, day: 1 }),
    );
    expect(options).toMatchSnapshot();
  });
});
