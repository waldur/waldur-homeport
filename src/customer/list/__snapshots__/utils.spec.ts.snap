// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeAccountingPeriods iterates through months 1`] = `
Array [
  Object {
    "label": "October, 2021",
    "value": Object {
      "current": true,
      "month": 10,
      "year": 2021,
    },
  },
  Object {
    "label": "September, 2021",
    "value": Object {
      "current": false,
      "month": 9,
      "year": 2021,
    },
  },
  Object {
    "label": "August, 2021",
    "value": Object {
      "current": false,
      "month": 8,
      "year": 2021,
    },
  },
  Object {
    "label": "July, 2021",
    "value": Object {
      "current": false,
      "month": 7,
      "year": 2021,
    },
  },
  Object {
    "label": "June, 2021",
    "value": Object {
      "current": false,
      "month": 6,
      "year": 2021,
    },
  },
  Object {
    "label": "May, 2021",
    "value": Object {
      "current": false,
      "month": 5,
      "year": 2021,
    },
  },
  Object {
    "label": "April, 2021",
    "value": Object {
      "current": false,
      "month": 4,
      "year": 2021,
    },
  },
  Object {
    "label": "March, 2021",
    "value": Object {
      "current": false,
      "month": 3,
      "year": 2021,
    },
  },
  Object {
    "label": "February, 2021",
    "value": Object {
      "current": false,
      "month": 2,
      "year": 2021,
    },
  },
  Object {
    "label": "January, 2021",
    "value": Object {
      "current": false,
      "month": 1,
      "year": 2021,
    },
  },
]
`;
